File: org/sathrek/sky_archipelago/worldgen/generator/structure/v2/common/placement/RelocatedPlacementContext.java
package org.sathrek.sky_archipelago.worldgen.generator.structure.v2.common.placement;

import net.minecraft.resources.ResourceKey;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.level.ChunkPos;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.levelgen.structure.BoundingBox;
import net.minecraft.world.level.levelgen.structure.StructurePiece;

public final class RelocatedPlacementContext {
   private final ResourceLocation structureId;
   private final ChunkPos startChunk;
   private final ResourceKey<Level> dimension;
   private final int authorityAnchorY;
   private final boolean yLockEnabled;
   private final PostProcessDriftTelemetry telemetry;
   private final PostProcessDriftPolicy policy;
   private final StructurePlacementAdapter adapter;

   RelocatedPlacementContext(
      ResourceLocation structureId,
      ChunkPos startChunk,
      ResourceKey<Level> dimension,
      int authorityAnchorY,
      boolean yLockEnabled,
      PostProcessDriftPolicy policy,
      StructurePlacementAdapter adapter
   ) {
      this.structureId = structureId;
      this.startChunk = startChunk;
      this.dimension = dimension;
      this.authorityAnchorY = authorityAnchorY;
      this.yLockEnabled = yLockEnabled;
      this.policy = policy;
      this.adapter = adapter;
      this.telemetry = new PostProcessDriftTelemetry();
   }

   public ResourceLocation structureId() {
      return this.structureId;
   }

   public ChunkPos startChunk() {
      return this.startChunk;
   }

   public ResourceKey<Level> dimension() {
      return this.dimension;
   }

   public int authorityAnchorY() {
      return this.authorityAnchorY;
   }

   public boolean yLockEnabled() {
      return this.yLockEnabled;
   }

   public boolean suppressIglooSurfaceAdjustment(StructurePiece piece) {
      return this.adapter.suppressIglooSurfaceAdjustment(this, piece);
   }

   public PostProcessDriftDecision evaluateDrift(StructurePiece piece, BoundingBox before, BoundingBox after) {
      return this.policy.evaluate(this, piece, before, after);
   }

   public void recordDrift(StructurePiece piece, BoundingBox before, BoundingBox after, PostProcessDriftDecision decision) {
      this.telemetry.record(piece, before, after, decision, this);
   }

   public String pieceRole(StructurePiece piece) {
      return this.adapter.pieceRole(piece);
   }
}

File: bedrock-addon/scripts/worldgen/pieces/RelocatedPlacementContext.js
export class RelocatedPlacementContext {
  constructor({structureId,startChunk,dimension,authorityAnchorY,yLockEnabled=true,policy,adapter,telemetry=null}={}){this.structureId=structureId;this.startChunk=startChunk;this.dimension=dimension;this.authorityAnchorY=authorityAnchorY;this.yLockEnabled=!!yLockEnabled;this.policy=policy;this.adapter=adapter;this.telemetry=telemetry;}
  pieceRole(piece){return this.adapter?.pieceRole?.(piece)??piece?.role??"unknown";}
  suppressIglooSurfaceAdjustment(piece){return !!this.adapter?.suppressIglooSurfaceAdjustment?.(this,piece);}
  evaluateDrift(piece,before,after){return this.policy?.evaluate?.(this,piece,before,after)??{correctionDy:0,pieceRole:this.pieceRole(piece),expectedDy:piece?.expectedDy??null,reason:"default_allow"};}
  recordDrift(piece,before,after,decision){this.telemetry?.record?.(piece,before,after,decision,this);}
}
export class PostProcessDriftDecision { static allow(pieceRole,expectedDy,reason){return{correctionDy:0,pieceRole,expectedDy,reason};}static clamp(correctionDy,pieceRole,expectedDy,reason){return{correctionDy,pieceRole,expectedDy,reason};} }
export class PostProcessDriftTelemetry {
 constructor(enabled=true){this.enabled=enabled;this.records=[];}
 record(piece,before,after,decision,context){if(!this.enabled||!before||!after)return;const drift={dx:after.min.x-before.min.x,dy:after.min.y-before.min.y,dz:after.min.z-before.min.z};const record={structureId:context?.structureId??"unknown",startChunk:context?.startChunk??null,piece:piece?.id??"unknown",role:decision?.pieceRole??piece?.role??"unknown",drift,expectedDy:decision?.expectedDy??null,correctionDy:decision?.correctionDy??0,reason:decision?.reason??"none",flagged:Math.abs(drift.dy)>4};this.records.push(record);if(record.flagged)console.warn(`[Sky Archipelago] structure piece drift: ${JSON.stringify(record)}`);}
}
